src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala

```diff
@@ -0,0 +1,229 @@
+
+  //Get a blob of unlaunched workflows, flip their status, and queue them for submission.
+  def getUnlaunchedWorkflowBatch()(implicit executionContext: ExecutionContext): Future[WorkflowSubmissionMessage] = {
+    val unlaunchedWfOptF = dataSource.inTransaction { dataAccess =>
+      //grab a bunch of unsubmitted workflows
+      dataAccess.workflowQuery.findUnsubmittedWorkflows().take(batchSize).result map { wfRecs =>
+        if (wfRecs.nonEmpty) wfRecs
+        else {
+          //they should also all have the same submission ID
+          val wfsWithASingleSubmission = wfRecs.filter(_.submissionId == wfRecs.head.submissionId)
+          //instead: on update, if we don't get the right number back, roll back the txn
+          //TODO: biden has an incoming change here, with optimistic locking
+          dataAccess.workflowQuery.batchUpdateStatus(wfsWithASingleSubmission.map(_.id), WorkflowStatuses.Launching)
+          wfsWithASingleSubmission
+        }
+      }
+    }
+
+    //if we find any, next step is to submit them. otherwise, look again.
+    unlaunchedWfOptF.map {
+      case workflowRecs if workflowRecs.nonEmpty => SubmitWorkflowBatch(workflowRecs.map(_.id))
+      case _ => LookForWorkflows
+    }
+  }
+
+  def buildWorkflowOpts(workspace: WorkspaceRecord, submissionId: UUID, user: RawlsUser, token: String, billingProject: RawlsBillingProject) = {
+    ExecutionServiceWorkflowOptions(
+      s"gs://${workspace.bucketName}/${submissionId}",
+      workspace.namespace,
+      user.userEmail.value,
+      token,
+      billingProject.cromwellAuthBucketUrl
+    )
+  }
+
+  def getWdl(methodConfig: MethodConfiguration): ReadWriteAction[String] = {
+    withMethod(methodConfig.methodRepoMethod.methodNamespace, methodConfig.methodRepoMethod.methodName, methodConfig.methodRepoMethod.methodVersion, getUserInfo(credential)) { method =>
+      withWdl(method) { wdl =>
+        DBIO.successful(wdl)
+      }
+    }
+  }
+
+  //submit the batch of workflows with the given ids
+  def submitWorkflowBatch(workflowIds: Seq[Long])(implicit executionContext: ExecutionContext): Future[WorkflowSubmissionMessage] = {
+
+    val workflowBatchFuture = dataSource.inTransaction { dataAccess =>
+      for {
+      //Load a bunch of things we'll need to reconstruct information:
+      //The list of workflows in this submission
+        wfRecs <- dataAccess.workflowQuery.findWorkflowByIds(workflowIds).result
+        submissionId = wfRecs.head.submissionId
+        _ = assert(wfRecs.forall(_.submissionId == submissionId)) //sanity check they're all in the same submission
+        unpackedWfOpts <- DBIO.sequence(wfRecs.map(dataAccess.workflowQuery.loadWorkflow)) //reify to real workflows
+        _ = assert(unpackedWfOpts.size == wfRecs.size) //there should be the right number of them
+        _ = assert(unpackedWfOpts.forall(_.isDefined)) //and they should all have been found
+        unpackedWfs = for {wfOpt <- unpackedWfOpts; wf <- wfOpt} yield {
```

Comment:
wrap the unpackedWfOpts and unpackedWfs shenanigans in a function. And it could probably be 1 line

DBIO.sequence(wfRecs.map(dataAccess.workflowQuery.loadWorkflow.map(_.getOrElse(throw...))))

I suggest an exception because it seems bad if unpackedWfs and wfRecs have different sizes